content/blog/debugging-valgrind-gdb.md

```diff
@@ -0,0 +1,190 @@
+<http://www.gnu.org/software/gdb/bugs/>.
+Find the GDB manual and other documentation resources online at:
+    <http://www.gnu.org/software/gdb/documentation/>.
+
+For help, type "help".
+Type "apropos word" to search for commands related to "word"...
+Reading symbols from /usr/lib/R/bin/exec/R...
+(No debugging symbols found in /usr/lib/R/bin/exec/R)
+(gdb) target remote | vgdb --pid=72681
+Remote debugging using | vgdb --pid=72681
+relaying data between gdb and process 72681
+warning: remote target does not support file transfer, attempting to access files from local filesystem.
+Reading symbols from /lib64/ld-linux-x86-64.so.2...
+Reading symbols from /usr/lib/debug//lib/x86_64-linux-gnu/ld-2.31.so...
+0x0000000004001100 in _start () from /lib64/ld-linux-x86-64.so.2
+(gdb) continue
+Continuing.
+```
+
+At this point, the program starts running in the first terminal window, eventually hitting the memory error from before and saying
+
+```
+==72681== Invalid read of size 4
+==72681==    at 0x14387DAB: void dust::filter::resample_weight<float>(std::vector<float, std::allocator<float> >::const_iterator, unsigned long, float, unsigned long, __gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >) (filter_tools.hpp:20)
+==72681==    by 0x1437D28A: dust::Dust<sirs>::resample(std::vector<float, std::allocator<float> > const&, std::vector<unsigned long, std::allocator<unsigned long> >&) (dust.hpp:420)
+==72681==    by 0x143802A3: std::vector<sirs::real_t, std::allocator<sirs::real_t> > dust::filter::filter<sirs>(dust::Dust<sirs>*, dust::filter::filter_state_host<sirs::real_t>&, bool, std::vector<unsigned long, std::allocator<unsigned long> >) (filter.hpp:59)
+==72681==    by 0x14376629: cpp11::writable::r_vector<double> dust::r::run_filter<sirs, dust::filter::filter_state_host<float> >(dust::Dust<sirs>*, cpp11::sexp&, cpp11::sexp&, std::vector<unsigned long, std::allocator<unsigned long> >&, bool) (interface.hpp:442)
+[...snip...]
+==72681==
+==72681== (action on error) vgdb me ...
+```
+
+Control was then passed back to the second window running gdb:
+
+```
+Program received signal SIGTRAP, Trace/breakpoint trap.
+dust::filter::resample_weight<float> (w=1.1479437e-41, n=100, u=0.999940574, offset=0, idx=99)
+    at /home/rfitzjoh/lib/R/library/dust/include/dust/filter_tools.hpp:20
+20	      ww += *w;
+(gdb)
+```
+
+and all the usual gdb tricks work:
+
+```
+(gdb) print u
+$1 = 0.999940574
+(gdb) print du
+$2 = 0.602285564
+(gdb) print uu
+$3 = 60.2285919
+(gdb) print ww
+$4 = 60.2285576
+```
+
+Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `ww` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)
```

Comment:
Do you mean this? Was confusing at first.

```suggestion
Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `w` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)
```

Reply:
Oh, just looked at the PR. It was `j`

```suggestion
Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `j` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)
```

Reply:
Ah, this is unclear. It's actually the `w` being iterated past its end which was raising the first error. Then `j` ends up with something that will break the index later and cascade through to cause a crash. I'll tidy this up tomorrow